deepchem/hyper/random_search.py

```diff
@@ -0,0 +1,284 @@
+  >>> import deepchem as dc
+  >>> optimizer = dc.hyper.RandomHyperparamOpt(lambda **p: dc.models.GraphConvModel(**p))
+
+  Here's a more sophisticated example that shows how to optimize only
+  some parameters of a model. In this case, we have some parameters we
+  want to optimize, and others which we don't. To handle this type of
+  search, we create a `model_builder` which hard codes some arguments
+  (in this case, `max_iter` is a hyperparameter which we don't want
+  to search over)
+
+  >>> import deepchem as dc
+  >>> import numpy as np
+  >>> from sklearn.linear_model import LogisticRegression as LR
+  >>> # generating data
+  >>> X = np.arange(1, 11, 1).reshape(-1, 1)
+  >>> y = np.hstack((np.zeros(5), np.ones(5)))
+  >>> dataset = dc.data.NumpyDataset(X, y)
+  >>> # splitting dataset into train and test
+  >>> splitter = dc.splits.RandomSplitter()
+  >>> train_dataset, test_dataset = splitter.train_test_split(dataset)
+  >>> # metric to evaluate result of a set of parameters
+  >>> metric = dc.metrics.Metric(dc.metrics.accuracy_score)
+  >>> # defining `model_builder`
+  >>> def model_builder(**model_params):
+  ...   penalty = model_params['penalty']
+  ...   solver = model_params['solver']
+  ...   lr = LR(penalty=penalty, solver=solver, max_iter=100)
+  ...   return dc.models.SklearnModel(lr)
+  >>> # the parameters which are to be optimized
+  >>> params = {
+  ...   'penalty': ['l1', 'l2'],
+  ...   'solver': ['liblinear', 'saga']
+  ...   }
+  >>> # Creating optimizer and searching over hyperparameters
+  >>> optimizer = dc.hyper.RandomHyperparamOpt(model_builder)
+  >>> best_model, best_hyperparams, all_results = \
+  optimizer.hyperparam_search(params, train_dataset, test_dataset, metric, 3)
+  >>> best_hyperparams  # the best hyperparameters found using random search
+  {'penalty': 'l2', 'solver': 'saga'}
+
+  """
+
+  def hyperparam_search(
+      self,
+      params_dict: Dict,
+      train_dataset: Dataset,
+      valid_dataset: Dataset,
+      metric: Metric,
+      n_search: int,
+      output_transformers: List[Transformer] = [],
+      nb_epoch: int = 10,
+      use_max: bool = True,
+      logdir: Optional[str] = None,
+      logfile: Optional[str] = None,
+      **kwargs,
+  ):
+    """Perform random hyperparams search according to `params_dict`.
+
+    Each key of the `params_dict` is a model_param. The 
+    values should either be a list of potential values of that hyperparam
+    or a Random Sampler from scipy.stats
+
+    Parameters
+    ----------
+    params_dict: Dict
+      Maps each hyperparameter name (string) to either a list of possible
+      parameter values or a Random Sampler from scipy.stats. If a random
+      sampler is passed, it must have the rvs method to sample values.
+    train_dataset: Dataset
+      dataset used for training
+    valid_dataset: Dataset
+      dataset used for validation (optimization on valid scores)
+    metric: Metric
+      metric used for evaluation
+    n_search: Int
```

Comment:
Can `n_search` be something like `n_perc`? By this, the user will specify what percentage of unique hyperparameter sets to use in search rather than the number of hyperparameter searches to make. This might also be useful since the user can assert a confidence level to the results produced.

Reply:
This is an intriguing thought. How would we handle continuous search ranges with `n_perc`? The code above assumes we have a Random Sampler to sample continuous ranges